### PR TITLE
Add support for pri files

### PR DESCRIPTION
**Pull request by cnbluefire**

### CODE REVIEW

**Comment by AndrewKeepCoding** [comment]
@cnbluefire Thanks for the PR! I guess I'll be able to take a look in a couple of days.

**Comment by AndrewKeepCoding** [comment]
@cnbluefire 
Sorry for have kept you waiting.

The WinUI3Localizer is trying to stick to the following:
> - Switch languages without app restarting
> - You/users can edit localized strings even after deployment
> - You/users can add new languages even after deployment
> - Use standard Resources.resw (see [Microsoft docs](https://learn.microsoft.com/en-us/windows/uwp/app-resources/localize-strings-ui-manifest))

and since PRI files can't be? (or are hard to?) add/edit, I'm not sure if PRI files support should be built into the WinUI3Localizer.

If you need to use PRI files, that can be done with resources features WinAppSDK provides. Or maybe, will you be open to the idea of a separate NuGet package?

Also, I'm struggling with the test you added. It seems that it breaks the test project and won't run any tests. Any tips?

**Comment by cnbluefire** [comment]
When using MSIX packaging, the installation directory is read-only, so modifying resw or pri files during runtime doesn't make much sense.

I implemented support for pri files as a reader, which simply reads resources from the pri file into a LanguageDictionary and does not affect any existing functionality, such as adding new languages to Localizer during runtime.

WinUI3 uses MRT core to handle pri files, and MRT core only implements methods to read pri files, not to modify them, so there is no simple way to modify pri files.

Are there any error messages when debugging the test project? In my environment, the test project runs normally.

**Comment by cnbluefire** [comment]
The Test project might not be running due to issues with the sln configuration. I have attempted to fix it.

**Comment by AndrewKeepCoding** [comment]
> When using MSIX packaging, the installation directory is read-only, so modifying resw or pri files during runtime doesn't make much sense.

The sample app uses the `LocalState` folder to store the localized strings as *.rews files and they can be modified after deployment.

```cs
 // Initialize a "Strings" folder in the "LocalFolder" for the packaged app.
 StorageFolder localFolder = ApplicationData.Current.LocalFolder;
 StorageFolder stringsFolder = await localFolder.CreateFolderAsync("Strings", CreationCollisionOption.OpenIfExists);
 StringsFolderPath = stringsFolder.Path;

// Create string resources file from app resources if doesn't exists.
await MakeSureStringResourceFileExists(stringsFolder, "en-US", "Resources.resw");
await MakeSureStringResourceFileExists(stringsFolder, "en-US", "ErrorMessages.resw");
await MakeSureStringResourceFileExists(stringsFolder, "es-ES", "Resources.resw");
await MakeSureStringResourceFileExists(stringsFolder, "es-ES", "ErrorMessages.resw");
await MakeSureStringResourceFileExists(stringsFolder, "ja", "Resources.resw");
await MakeSureStringResourceFileExists(stringsFolder, "ja", "ErrorMessages.resw");
```

I still can't see why you want to the WinUI3Localizer to support *.pri files when they are supported by default on WinAppSDK apps. Are you having some issues with the default *.pri support?🤔

**Comment by AndrewKeepCoding** [comment]
> Are there any error messages when debugging the test project? In my environment, the test project runs normally.

The tests are still not running. I'm swamped now and didn't take a look why it's not working but these are the errors:
```
Could not determine target device configuration. Exception: System.Runtime.InteropServices.COMException (0x80020009): The method or operation is not implemented.
   at EnvDTE.Property.get_Value()
   at Microsoft.VisualStudio.TestWindow.ShellServices.ProjectDataImpl.<GetRemoteMachineAddressAsync>d__36.MoveNext()
========== Starting test run ==========
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.5+1caef2f33e (64-bit .NET 6.0.28)
[xUnit.net 00:00:01.11]   Starting:    WinUI3Localizer.Tests
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.Runtime.InteropServices.COMException (0x80070491): There was no match for the specified key in the index.
   at Microsoft.Windows.ApplicationModel.DynamicDependency.NativeMethods.MddBootstrapInitialize2_Throw(UInt32 majorMinorVersion, String versionTag, PackageVersion packageVersion, InitializeOptions options)
   at Microsoft.Windows.ApplicationModel.DynamicDependency.Bootstrap.Initialize(UInt32 majorMinorVersion, String versionTag, PackageVersion minVersion)
   at Microsoft.Windows.ApplicationModel.DynamicDependency.Bootstrap.Initialize(UInt32 majorMinorVersion, String versionTag)
   at Microsoft.Windows.ApplicationModel.DynamicDependency.Bootstrap.Initialize(UInt32 majorMinorVersion)
   at WinUI3Localizer.Tests.PriResourceReaderTest.<>c__DisplayClass0_0.<GetItems_ReturnsAllItems>b__0() in C:\Projects\AndrewKeepCoding\WinUI3Localizer-cnbluefire\WinUI3Localizer.Tests\PriResourceReaderTest.cs:line 20
   at System.Threading.Thread.StartCallback()

========== Test run aborted: 0 Tests (0 Passed, 0 Failed, 0 Skipped) run in < 1 ms ==========
```

**Comment by cnbluefire** [comment]
Even with the use of pri files, it is possible to save modifications to resw during runtime. Loading resources from both the pri file and resw resources does not conflict. This PR simply adds a way to retrieve resources from the pri file.

WinUI3 natively supports PRI files, but content set using x:Uid requires restarting the app to correctly display after changing languages. I want to use WinUI3Localizer to respond to language changes during runtime. If you don't want to merge, I can create a nuget package that depends on WinUI3Localizer.

Running the test project requires installing the WinAppSDK 1.4 runtime locally, which can be downloaded from [here](https://learn.microsoft.com/zh-cn/windows/apps/windows-app-sdk/downloads).

**Comment by AndrewKeepCoding** [comment]
> I want to use WinUI3Localizer to respond to language changes during runtime. 

Convinced. This is a fair case. Sorry I didn't see it earlier. 😵

I'll try again with the tests tonight.

**Comment by AndrewKeepCoding** [comment]
After installing WinAppSDK v1.4.5 (1.4.240211001) runtime, at least the other tests are running, but the Pri files test throws:

`System.Runtime.InteropServices.COMException: 'There was no match for the specified key in the index.`

at

```cs
Microsoft.Windows.ApplicationModel.DynamicDependency.Bootstrap.Initialize(0x00010004);
```

I'm considering dropping the test for the moment.

**Comment by cnbluefire** [comment]
I'll temporarily use my own fork, and hope to discuss this PR when you have some free time.

File: WinUI3Localizer.Tests/PriResourceReaderTest.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+{
+    [Fact]
+    public void GetItems_ReturnsAllItems()
+    {
+        LanguageDictionary.Item[]? resourcesItems = null;
+        LanguageDictionary.Item[]? errorMessagesItems = null;
+
+        Thread? thread = new Thread(() =>
+        {
+            Microsoft.Windows.ApplicationModel.DynamicDependency.Bootstrap.Initialize(0x00010004);
```
Comment (AndrewKeepCoding) [inline WinUI3Localizer.Tests/PriResourceReaderTest.cs]: Can you change the version to 0x00010005?

File: WinUI3Localizer.Tests/WinUI3Localizer.Tests.csproj [hunk 2]
```diff
@@ -5,11 +5,14 @@
...
     <Nullable>enable</Nullable>
     <IsPackable>false</IsPackable>
+    <Platforms>x86;x64;arm64</Platforms>
+    <RuntimeIdentifiers>win10-x86;win10-x64;win10-arm64</RuntimeIdentifiers>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="FluentAssertions" Version="6.9.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.4.1" />
+    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.230822000" />
```
Comment (AndrewKeepCoding) [inline WinUI3Localizer.Tests/WinUI3Localizer.Tests.csproj]: Can you change the version to "1.5.240404000"?

**Comment by AndrewKeepCoding** [comment]
It worked with the latest WinAppSDK. My problem might be due to some cached packages.
Can you apply the 2 changes I left above?

**Comment by AndrewKeepCoding** [comment]
@cnbluefire Published it as v2.2.0. Thanks for the contribution!

**Pull request merged**